File: algo/src/main/java/com/hartwig/actin/algo/evaluation/othercondition/HasPotentialAbsorptionDifficulties.java
package com.hartwig.actin.algo.evaluation.othercondition;

import java.util.Set;

import com.google.common.collect.Sets;
import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.doid.DoidConstants;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.algo.evaluation.util.Format;
import com.hartwig.actin.algo.othercondition.OtherConditionSelector;
import com.hartwig.actin.clinical.datamodel.Complication;
import com.hartwig.actin.clinical.datamodel.PriorOtherCondition;
import com.hartwig.actin.clinical.datamodel.Toxicity;
import com.hartwig.actin.clinical.datamodel.ToxicitySource;
import com.hartwig.actin.doid.DoidModel;

import org.jetbrains.annotations.NotNull;

public class HasPotentialAbsorptionDifficulties implements EvaluationFunction {

    static final String GASTROINTESTINAL_DISORDER_CATEGORY = "gastrointestinal disorder";

    static final Set<String> TOXICITIES_CAUSING_ABSORPTION_DIFFICULTY = Sets.newHashSet();

    static {
        TOXICITIES_CAUSING_ABSORPTION_DIFFICULTY.add("diarrhea");
        TOXICITIES_CAUSING_ABSORPTION_DIFFICULTY.add("nausea");
        TOXICITIES_CAUSING_ABSORPTION_DIFFICULTY.add("vomit");
    }

    @NotNull
    private final DoidModel doidModel;

    HasPotentialAbsorptionDifficulties(@NotNull final DoidModel doidModel) {
        this.doidModel = doidModel;
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        Set<String> conditions = Sets.newHashSet();
        for (PriorOtherCondition condition : OtherConditionSelector.selectClinicallyRelevant(record.clinical().priorOtherConditions())) {
            for (String doid : condition.doids()) {
                if (doidModel.doidWithParents(doid).contains(DoidConstants.GASTROINTESTINAL_SYSTEM_DISEASE_DOID)) {
                    conditions.add(doidModel.resolveTermForDoid(doid));
                }
            }
        }

        if (!conditions.isEmpty()) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.PASS)
                    .addPassSpecificMessages("Patient has potential absorption difficulties due to " + Format.concat(conditions))
                    .addPassGeneralMessages("Potential absorption difficulties: " + Format.concat(conditions))
                    .build();
        }

        Set<String> complications = Sets.newHashSet();
        if (record.clinical().complications() != null) {
            for (Complication complication : record.clinical().complications()) {
                if (isOfCategory(complication, GASTROINTESTINAL_DISORDER_CATEGORY)) {
                    complications.add(complication.name());
                }
            }
        }

        if (!complications.isEmpty()) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.PASS)
                    .addPassSpecificMessages("Patient has potential absorption difficulties due to " + Format.concat(complications))
                    .addPassGeneralMessages("Potential absorption difficulties: " + Format.concat(complications))
                    .build();
        }

        Set<String> toxicities = Sets.newHashSet();
        for (Toxicity toxicity : record.clinical().toxicities()) {
            if (toxicity.source() == ToxicitySource.QUESTIONNAIRE || (toxicity.grade() != null && toxicity.grade() >= 2)) {
                for (String termToFind : TOXICITIES_CAUSING_ABSORPTION_DIFFICULTY) {
                    if (toxicity.name().toLowerCase().contains(termToFind.toLowerCase())) {
                        toxicities.add(toxicity.name());
                    }
                }
            }
        }

        if (!toxicities.isEmpty()) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.PASS)
                    .addPassSpecificMessages("Patient has potential absorption difficulties due to " + Format.concat(toxicities))
                    .addPassGeneralMessages("Potential absorption difficulties: " + Format.concat(toxicities))
                    .build();
        }

        return EvaluationFactory.unrecoverable()
                .result(EvaluationResult.FAIL)
                .addFailSpecificMessages("No potential reasons for absorption problems identified")
                .addFailGeneralMessages("No potential absorption difficulties identified")
                .build();
    }

    private static boolean isOfCategory(@NotNull Complication complication, @NotNull String categoryToFind) {
        for (String category : complication.categories()) {
            if (category.toLowerCase().contains(categoryToFind.toLowerCase())) {
                return true;
            }
        }
        return false;
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/othercondition/HasPotentialAbsorptionDifficulties.kt
package com.hartwig.actin.algo.evaluation.othercondition

import com.hartwig.actin.algo.doid.DoidConstants
import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.algo.evaluation.util.Format.concat
import com.hartwig.actin.algo.evaluation.util.ValueComparison.stringCaseInsensitivelyMatchesQueryCollection
import com.hartwig.actin.algo.othercondition.OtherConditionSelector
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.clinical.Complication
import com.hartwig.actin.datamodel.clinical.ToxicitySource
import com.hartwig.actin.doid.DoidModel

class HasPotentialAbsorptionDifficulties(private val doidModel: DoidModel) : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        val conditions = OtherConditionSelector.selectClinicallyRelevant(record.priorOtherConditions).flatMap { it.doids }
            .filter { doidModel.doidWithParents(it).any { doid -> doid in DoidConstants.ABSORPTION_DIFFICULTIES_DOID_SET } }
            .map { doidModel.resolveTermForDoid(it) }

        if (conditions.isNotEmpty()) {
            return EvaluationFactory.pass("Potential absorption difficulties: " + concat(conditions.filterNotNull()))
        }
        val complications = record.complications?.filter { isOfCategory(it, GASTROINTESTINAL_DISORDER_CATEGORY) }
            ?.map { it.name } ?: emptyList()

        if (complications.isNotEmpty()) {
            return EvaluationFactory.pass("Potential absorption difficulties: " + concat(complications))
        }
        val toxicities = record.toxicities
            .filter { it.source == ToxicitySource.QUESTIONNAIRE || (it.grade ?: 0) >= 2 }
            .map { it.name }
            .filter { stringCaseInsensitivelyMatchesQueryCollection(it, TOXICITIES_CAUSING_ABSORPTION_DIFFICULTY) }

        return if (toxicities.isNotEmpty()) {
            EvaluationFactory.pass("Potential absorption difficulties: " + concat(toxicities))
        } else
            EvaluationFactory.fail("No potential reasons for absorption problems identified")
    }

    companion object {
        const val GASTROINTESTINAL_DISORDER_CATEGORY: String = "gastrointestinal disorder"
        val TOXICITIES_CAUSING_ABSORPTION_DIFFICULTY = setOf("diarrhea", "nausea", "vomit")

        private fun isOfCategory(complication: Complication, categoryToFind: String): Boolean {
            return complication.categories.any { it.lowercase().contains(categoryToFind.lowercase()) }
        }
    }
}